### PR TITLE
feat: Add support for TypeScript key query (`keyof X`) types

### PR DESCRIPTION
**Pull request by ExE-Boss**
Part of #50

This implements parsing and handling of TypeScript’s key query types, which are somewhat common.

---

review?(@sandersn)

### CODE REVIEW

File: lib/NodeType.js [hunk 1]
```diff
@@ -26,9 +26,10 @@ const NodeType = {
...
   PARENTHESIS: 'PARENTHESIS',
   TYPE_QUERY: 'TYPE_QUERY',
+  KEY_QUERY: 'KEY_QUERY',
   IMPORT: 'IMPORT',
 };
 
 /** @typedef {keyof typeof NodeType} Type */
 
-module.exports = NodeType;
+module.exports = /** @type {{readonly [T in Type]: T}} */ (NodeType);
```
Comment (sandersn) [inline lib/NodeType.js]: that type seems circular (`NodeType -> Type -> NodeType`). Do you start getting weird implicit-any or circularity errors if you turn on `"strict": true` in tsconfig?

File: tests/test_parsing.js [hunk 2]
```diff
@@ -1388,6 +1422,10 @@ describe('Parser', function() {
 });
 
 
+/**
+ * @template T
```
Comment (sandersn) [inline tests/test_parsing.js]: isn't `T=string` always true?

File: tests/test_parsing.js [hunk 3]
```diff
@@ -1561,3 +1599,26 @@ function createParenthesizedNode(value) {
     value: value,
   };
 }
+
+/**
+ * @template T
```
Comment (sandersn) [inline tests/test_parsing.js]: same. I think that `T={type: string, name: string }` (or maybe a more specific type)

Comment (sandersn) [reply tests/test_parsing.js]: anyway, this is test code, so I don't know if the overhead of jsdoc types is worth it here.

File: lib/NodeType.js (hunk 1 above)
Comment (ExE-Boss) [reply lib/NodeType.js]: This is not circular, as this doesn’t affect the definition of the `NodeType` object, only the `NodeType.js` file export.

File: tests/test_parsing.js (hunk 3 above)
Comment (ExE-Boss) [reply tests/test_parsing.js]: Well, in&nbsp;this&nbsp;case, `T`&nbsp;can&nbsp;be&nbsp;any&nbsp;subtype of&nbsp;`Node`.

File: lib/NodeType.js (hunk 1 above)
Comment (sandersn) [reply lib/NodeType.js]: Oh, right, never mind. I was thinking NodeType was an *alias* for `module.exports`, but it's not.

I've spent too much of my life trying to check code like this:

```js
module.exports = NodeType = {
  ...
}
NodeType.x = 1
NodeType.y = 2
// etc
```

File: tests/test_parsing.js (hunk 3 above)
Comment (sandersn) [reply tests/test_parsing.js]: In that case I think it should be `@template {Node} T` -- as is `T` will bind to any type, so you could get a return type `{ type: NodeType, name: string | boolean | undefined }` or something crazy like that.

**Comment by ExE-Boss** [comment]
So, can this be merged now? (I don’t have write access to this repository)

**Comment by sandersn** [comment]
Oops! Sorry about that. You should have access now.

**Comment by sandersn** [comment]
Unfortunately I don't have time to publish this afternoon. Let me know if you need this before Monday and I can publish it from my laptop.

**Pull request merged**